### PR TITLE
Muddle V2

### PR DESCRIPTION
**Pull request by ejfitzgerald**

### CODE REVIEW

File: apps/beacon/main.cpp [hunk 1]
```diff
@@ -115,6 +114,7 @@ int main()
   }
   std::this_thread::sleep_for(std::chrono::milliseconds(500));
 
+#if 0
```
Comment (ejfitzgerald) [inline apps/beacon/main.cpp]: @n-hutton lets have a chat about this tomorrow

Comment (troels-im) [reply apps/beacon/main.cpp]: I'd like to be in that discussion.

File: apps/constellation/constellation.cpp [hunk 2]
```diff
@@ -533,19 +543,15 @@ void Constellation::Run(UriList const &initial_peers, core::WeakRunnable bootstr
   FETCH_LOG_INFO(LOGGING_NAME, "Shutting down...");
 
   http_.Stop();
-  p2p_.Stop();
```
Comment (troels-im) [inline apps/constellation/constellation.cpp:516]: P2P service is now an intrinsic part of muddle, I guess? Nice 👍

File: apps/muddle-node/main.cpp [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+struct AggregateData
+{
+  MsgCounters counters{};
+  std::size_t total_messages{0};
+};
+
+using Statistics = fetch::SynchronisedState<AggregateData>;
+
+std::atomic<bool>        gActive{true};
+std::atomic<std::size_t> gInterruptCount{0};
```
Comment (troelsfr) [inline apps/muddle-node/main.cpp]: Style?

File: apps/muddle-node/main.cpp [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  // signal that the program should stop
+  gActive = false;
+
+  if (interrupt_count >= 3)
+  {
+    std::exit(1);
+  }
+}
+
+ProverPtr RestoreOrCreateKey(ParamsParser const &params)
```
Comment (troels-im) [inline apps/muddle-node/main.cpp:102]: It feels like this is something that would be useful to have some where in the lib itself such that other apps can use it.

File: libs/core/include/core/serializers/exception.hpp [hunk 5]
```diff
@@ -54,6 +57,10 @@ class SerializableException : public std::exception
 
   void StackTrace() const;
```
Comment (troels-im) [inline libs/core/include/core/serializers/exception.hpp:58]: If the ticket does not already exist, let's make one to deprecate StackTrace?

File: libs/core/src/periodic_runnable.cpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+namespace fetch {
+namespace core {
+
+PeriodicRunnable::PeriodicRunnable(Duration const &period)
+  : last_executed_{Clock::now()}
+  , interval_{period}
+{
+}
+
+bool PeriodicRunnable::IsReadyToExecute() const
```
Comment (troels-im) [inline libs/core/src/periodic_runnable.cpp:29]: We might want to make runnable sortable according to next expected runtime and put time in a priority queue inside reactor.

File: libs/http/include/http/module.hpp [hunk 7]
```diff
@@ -53,6 +53,7 @@ class HTTPModule
   HTTPModule(HTTPModule &&rhs)      = delete;
   HTTPModule &operator=(HTTPModule const &rhs) = delete;
   HTTPModule &operator=(HTTPModule &&rhs) = delete;
+  virtual ~HTTPModule() = default;
```
Comment (troels-im) [inline libs/http/include/http/module.hpp]: :+1:

File: libs/dkg/src/pre_dkg_sync.cpp [hunk 8]
```diff
@@ -104,19 +104,22 @@ void PreDkgSync::Connect()
...
   for (auto const &peer : peers_)
   {
-    auto connections = muddle_.AsEndpoint().GetDirectlyConnectedPeers();
-    if (peer.first != muddle_.identity().identifier() &&
+    auto connections = muddle_.GetDirectlyConnectedPeers();
+    if (peer.first != muddle_.GetLocalAddress() &&
         std::find(connections.begin(), connections.end(), peer.first) == connections.end())
     {
+      // TODO(EJF): This pre sync needs a little bit of work
+#if 0
```
Comment (troels-im) [inline libs/dkg/src/pre_dkg_sync.cpp]: To be fixed?

File: libs/dkg/tests/gtests/dkg_tests.cpp [hunk 9]
```diff
@@ -470,7 +469,7 @@ void GenerateTest(uint32_t cabinet_size, uint32_t threshold, uint32_t qual_size,
     }
     if (ii >= (cabinet_size - qual_size))
     {
-      expected_qual.insert(committee[ii]->muddle.identity().identifier());
+      expected_qual.insert(committee[ii]->muddle->GetAddress());
```
Comment (troels-im) [inline libs/dkg/tests/gtests/dkg_tests.cpp:473]: What is the difference between `GetLocalAddress` and `GetAddress`?

File: libs/ledger/include/ledger/storage_unit/lane_controller_protocol.hpp [hunk 10]
```diff
@@ -17,28 +17,23 @@
...
 public:
   enum
   {
-    SHUTDOWN = 1,
-    START_SYNC,
-    STOP_SYNC,
-    INCOMING_PEERS,
-    OUTGOING_PEERS,
-    USE_THESE_PEERS,
+    USE_THESE_PEERS = 1
```
Comment (troels-im) [inline libs/ledger/include/ledger/storage_unit/lane_controller_protocol.hpp:32]: :+1:

File: libs/ledger/include/ledger/storage_unit/lane_service.hpp [hunk 11]
```diff
@@ -76,7 +73,7 @@ class LaneService
 
   // TODO(issue 7): Make config JSON
```
Comment (troels-im) [inline libs/ledger/include/ledger/storage_unit/lane_service.hpp]: Seems irrelevant at this point.

File: libs/ledger/src/shards/service_identifier.cpp [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  case ServiceIdentifier::Type::INVALID:
+    text = "Invalid";
+    break;
+  case ServiceIdentifier::Type::CORE:
+    text = "Core";
+    break;
+  case ServiceIdentifier::Type::HTTP:
+    text = "Http";
+    break;
+  case ServiceIdentifier::Type::DKG:
```
Comment (troels-im) [inline libs/ledger/src/shards/service_identifier.cpp:41]: Could we change the DKG identifier to BEACON or alternatively call it DRB? The DKG is just a subcomponent of the BEACON.

File: libs/ledger/src/storage_unit/lane_service.cpp [hunk 13]
```diff
@@ -57,28 +55,24 @@ std::string GeneratePrefix(std::string const &storage_path, uint32_t lane)
...
 }  // namespace
 
-LaneService::LaneService(NetworkManager nm, ShardConfig config, bool sign_packets, Mode mode)
+LaneService::LaneService(NetworkManager const &nm, ShardConfig config, Mode mode)
   : tx_store_(std::make_shared<TxStore>(meta::Log2(config.num_lanes)))
   , reactor_("LaneServiceReactor")
   , cfg_{std::move(config)}
 {
   // External Muddle network and RPC server
+  // TODO(EJF): !!! Must be set
```
Comment (troels-im) [inline libs/ledger/src/storage_unit/lane_service.cpp]: Something to sort for this PR?

File: libs/muddle/include/muddle/muddle_interface.hpp [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+   * @return The current network
+   */
+  virtual NetworkId const &GetNetwork() const = 0;
+
+  /**
+   * Get the address of this muddle node
+   *
+   * @return The address of the node
+   */
+  virtual Address const &GetAddress() const = 0;
```
Comment (troels-im) [inline libs/muddle/include/muddle/muddle_interface.hpp:107]: It seems that `GetAdress` and `GetLocalAddress` does the same thing. I'd suggest to streamline the names.

File: libs/muddle/include/muddle/muddle_interface.hpp [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+   */
+  virtual std::size_t GetNumDirectlyConnectedPeers() const = 0;
+
+  /**
+   * Determine if we are directly connected to the specified address
+   *
+   * @param address The address to check
+   * @return true if directly connected, otherwise false
+   */
+  virtual bool IsDirectlyConnected(Address const &address) const= 0;
```
Comment (troels-im) [inline libs/muddle/include/muddle/muddle_interface.hpp]: Let's think about how hard it would be to make a `IsReachable` function as well?

File: libs/network/include/network/service/protocol.hpp [hunk 16]
```diff
@@ -161,106 +142,6 @@ class Protocol
   virtual void ConnectionDropped(connection_handle_type /*connection_handle*/)
```
Comment (troels-im) [inline libs/network/include/network/service/protocol.hpp:142]: Do we still have connection handles and would it not be more appropriate to use context/address?

**Review by troels-im — APPROVED** [review]
This has a lot of good stuff.

File: apps/muddle-node/main.cpp (hunk 4 above)
Comment (ejfitzgerald) [reply apps/muddle-node/main.cpp:102]: Good suggestion -  I have another update in the pipeline will add it to the list

File: libs/core/include/core/serializers/exception.hpp (hunk 5 above)
Comment (ejfitzgerald) [reply libs/core/include/core/serializers/exception.hpp:58]: 👍

File: libs/core/src/periodic_runnable.cpp (hunk 6 above)
Comment (ejfitzgerald) [reply libs/core/src/periodic_runnable.cpp:29]: Yes this would require an interface update but would be a good idea. I am inclined to see if we run into any issues because of the ordering of the jobs in the reactor. For the most part we only use it for cases where the deadline is strictly a hint rather than a hard request.

File: libs/dkg/src/pre_dkg_sync.cpp (hunk 8 above)
Comment (ejfitzgerald) [reply libs/dkg/src/pre_dkg_sync.cpp]: Fixed

File: libs/ledger/src/storage_unit/lane_service.cpp (hunk 13 above)
Comment (ejfitzgerald) [reply libs/ledger/src/storage_unit/lane_service.cpp]: Fixed

File: libs/muddle/include/muddle/muddle_interface.hpp (hunk 14 above)
Comment (ejfitzgerald) [reply libs/muddle/include/muddle/muddle_interface.hpp:107]: done

File: libs/muddle/include/muddle/muddle_interface.hpp (hunk 15 above)
Comment (ejfitzgerald) [reply libs/muddle/include/muddle/muddle_interface.hpp]: It would be reasonably easy but might be quite unreliable

File: libs/network/include/network/service/protocol.hpp (hunk 16 above)
Comment (ejfitzgerald) [reply libs/network/include/network/service/protocol.hpp:142]: Nope we still have connection handles. This class needs a little more work

File: apps/constellation/constellation.cpp [hunk 17]
```diff
@@ -361,7 +384,15 @@ void Constellation::Run(UriList const &initial_peers, core::WeakRunnable bootstr
   http_open_api_module_->Reset(&http_);
   network_manager_.Start();
   http_network_manager_.Start();
-  muddle_.Start({p2p_port_});
+
+  // TODO(EJF): Tidy up
```
Comment (n-hutton) [inline apps/constellation/constellation.cpp:388]: errant todo?

File: apps/constellation/constellation.hpp [hunk 18]
```diff
@@ -139,18 +139,20 @@ class Constellation : public ledger::BlockSinkInterface
   using HttpModulePtr          = std::shared_ptr<HttpModule>;
   using HttpModules            = std::vector<HttpModulePtr>;
   using TransactionProcessor   = ledger::TransactionProcessor;
-  using TrustSystem            = p2p::P2PTrustBayRank<Muddle::Address>;
+  using TrustSystem            = p2p::P2PTrustBayRank<muddle::Address>;
```
Comment (n-hutton) [inline apps/constellation/constellation.hpp:142]: delete?

File: apps/muddle-node/main.cpp [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+
+struct AggregateData
+{
+  MsgCounters counters{};
+  std::size_t total_messages{0};
+};
+
+using Statistics = fetch::SynchronisedState<AggregateData>;
+
+std::atomic<bool>        gActive{true};
```
Comment (n-hutton) [inline apps/muddle-node/main.cpp:68]: non standard variable naming

File: libs/core/include/core/periodic_functor.hpp [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+//------------------------------------------------------------------------------
+
+#include "core/periodic_runnable.hpp"
+
+#include <functional>
+
+namespace fetch {
+namespace core {
+
+class PeriodicFunctor : public PeriodicRunnable
```
Comment (n-hutton) [inline libs/core/include/core/periodic_functor.hpp:27]: Where is this used? Could do with a class level comment?

File: libs/ledger/include/ledger/shards/service_identifier.hpp [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+#include "core/serializers/group_definitions.hpp"
+
+#include <cstdint>
+#include <string>
+
+namespace fetch {
+namespace ledger {
+
+class ServiceIdentifier
```
Comment (n-hutton) [inline libs/ledger/include/ledger/shards/service_identifier.hpp:28]: how is this used?

File: libs/muddle/include/muddle/muddle_endpoint.hpp [hunk 22]
```diff
@@ -67,6 +93,19 @@ class MuddleEndpoint
   virtual void Send(Address const &address, uint16_t service, uint16_t channel,
                     uint16_t message_num, Payload const &payload) = 0;
 
+  /**
```
Comment (n-hutton) [inline libs/muddle/include/muddle/muddle_endpoint.hpp:96]: 👍

File: libs/muddle/internal/muddle.hpp [hunk 23]
```diff
@@ -104,20 +111,20 @@ class MuddleEndpoint;
  *                         Underlying Network Connections
  *
  */
-class Muddle
+class Muddle : public MuddleInterface, public std::enable_shared_from_this<Muddle>
```
Comment (n-hutton) [inline libs/muddle/internal/muddle.hpp:114]: curious: when is muddle pointing to itself?

File: scripts/end_to_end_test/end_to_end_test.yaml [hunk 24]
```diff
@@ -121,7 +121,7 @@ test_description: >
 setup_conditions:
     test_name : "basic_multinode_POS"
     number_of_nodes: 3
-    node_connections: [[0, 1], [0, 2], [1, 0], [1, 2], [2, 0], [2, 1]]
+    node_connections: [[0, 1], [0, 2], [1, 2]]
```
Comment (n-hutton) [inline scripts/end_to_end_test/end_to_end_test.yaml:124]: 👍

**Review by n-hutton — APPROVED** [review]
LGTM - too much code to review in detail of course but I can't see anything suspect

File: libs/muddle/internal/muddle.hpp (hunk 23 above)
Comment (ejfitzgerald) [reply libs/muddle/internal/muddle.hpp:114]: When it registers itself with the registry. This is useful for collecting process wide stats on muddle instance.

File: libs/ledger/include/ledger/shards/service_identifier.hpp (hunk 21 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/shards/service_identifier.hpp:28]: It is the key in the manifest table used to identity a services like a lane. In the future we might refactor this subsystem a little. This is mostly a port of the code from the old p2p subsystem.

File: apps/constellation/constellation.hpp (hunk 18 above)
Comment (ejfitzgerald) [reply apps/constellation/constellation.hpp:142]: Keep, we need to update in the interface and also do more testing on this subsystem

File: apps/constellation/constellation.cpp (hunk 17 above)
Comment (ejfitzgerald) [reply apps/constellation/constellation.cpp:388]: Will tidy on following PR

File: apps/muddle-node/main.cpp (hunk 19 above)
Comment (ejfitzgerald) [reply apps/muddle-node/main.cpp:68]: Will tidy on following PR

File: libs/core/include/core/periodic_functor.hpp (hunk 20 above)
Comment (ejfitzgerald) [reply libs/core/include/core/periodic_functor.hpp:27]: Will tidy on following PR
